compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3407,25 +3407,26 @@ class Typer extends Namer
         case _ =>
       }
 
-      val approximation = Inferencing.approximateGADT(wtp)
+      // try GADT approximation, but only if we're trying to select a member
+      // Member lookup cannot take GADTs into account b/c of cache, so we
+      // approximate types based on GADT constraints instead. For an example,
+      // see MemberHealing in gadt-approximation-interaction.scala.
+      lazy val gadtApprox = Inferencing.approximateGADT(wtp)
```

Comment:
I think a lazy val is overkill here, instead I would do something like:
```scala
pt match {
  case pt: SelectionProto if ctx.gadt.nonEmpty =>
    val gadtApprox = Inferencing.approximateGADT(wtp)
    gadts.println(...)
    if (pt.isMatchedBy(gadtApprox))
      return tpd.Typed(tree, TypeTree(gadtApprox))
  case _ =>
}
```

Reply:
That doesn't mesh well with the debug statement, which would be best printed both when `pt` is and isn't a `SelectionProto`.

Reply:
If the concern is about performance, I seriously would not care about it unless it shows up in the performance test.

Reply:
No, I'm more concerned by clarity than peformance here, also I'm not convinced that we should print anything when pt is not a selectionproto: if I turn on the GADT printer, I want to see what the GADT logic has been doing, but I don't want my output full of messages about situation where it didn't do anything